### PR TITLE
hsmd: add preapprove invoice and keysend messages

### PR DESCRIPTION
**Pull request by ksedgwic**
These messages request payment approval from the hsmd for bolt11 and keysend payments:
- default implementation automatically approves all payments
- remote signer might not approve based on velocity controls, allowlist controls, manual approval, ...
- remote signer will not allow unapproved payments to be made (will refuse to sign commitment)
- declining a preapproval is much cleaner from a UX standpoint than preventing payment later

### CODE REVIEW

File: plugins/keysend.c
Comment (ksedgwic) [reply plugins/keysend.c:160]: [I thought I already commented, not sure where it went ...]

So `keysend.c` has the `End of ...` comments, but `libplugin-pay.c` doesn't seem to.  So we are "locally consistent", but happy to change in either direction ...

**Comment by cdecker** [comment]
Thanks @ksedgwic, just a couple of questions for my own understanding:

 - I'm guessing the pre-approval is there in order to avoid situations where `lightningd` would go ahead and attempt to `pay` or `keysend`, but then the signer would run into a policy violation, refusing to sign. This would then end up in the requesting daemon to hang indefinitely (restart seems like the only option here). Is that interpretation correct?
 - Would pre-approval result in any state changes on the signer? I.e., does the signer remember it approved a send, and would consider that for other decisions as well? Is this persisted?
 - How is concurrency handled? If I have two `pay` commands, that individually would be ok, but not together (velocity controls for example), how is the following sequence resolved:
   - Preapproval `A`
   - Preapproval `B`
   - Signature requests for `A`
   - Signature requests for `B`
 - If we track approvals persistently, how are these cleared again? Take for example a `pay` that was approved, but ultimately fails to complete, when is the approval cleared in order to allow future approvals to tap into the reservation for the failed `pay`?

On a more design-related thing: I'm assuming that `keysend` is the reason why we expose the pre-approval on the JSON-RPC. It being a plugin, it can't ask the signer to pre-approve, is that correct?

I'll review the code as soon as possible, but thought I needed a bit of a wider picture of how things are intended to work on a conceptual level 🙂

**Comment by devrandom** [comment]
> Thanks @ksedgwic, just a couple of questions for my own understanding:
> 
>     * I'm guessing the pre-approval is there in order to avoid situations where `lightningd` would go ahead and attempt to `pay` or `keysend`, but then the signer would run into a policy violation, refusing to sign. This would then end up in the requesting daemon to hang indefinitely (restart seems like the only option here). Is that interpretation correct?

Right, the other option is to return an error, but that would crash or at least restart the node.

There is another reason.  Seeing the invoice allows us to bind HTLCs (the payment hash) to payees (the entity who signed the invoice), which in turn allows payee to be allowlisted.

> 
>     * Would pre-approval result in any state changes on the signer? I.e., does the signer remember it approved a send, and would consider that for other decisions as well? Is this persisted?

Yes, the signer would start doing the following:

- track the payment hash, the amount approved and the amount actually sent
- adjust any velocity control state to account for the **amount approved**
- prevent sending more than was approved

> 
>     * How is concurrency handled? If I have two `pay` commands, that individually would be ok, but not together (velocity controls for example), how is the following sequence resolved:
>       
>       * Preapproval `A`
>       * Preapproval `B`
>       * Signature requests for `A`
>       * Signature requests for `B`

The signer would serialize the preapprovals internally, so the later request for preapproval would fail if the velocity is exceeded.  Note that velocity measures amount approved rather than amount actually sent, so that velocity controls can only fail a payment early - in the preapproval stage.

> 
>     * If we track approvals persistently, how are these cleared again? Take for example a `pay` that was approved, but ultimately fails to complete, when is the approval cleared in order to allow future approvals to tap into the reservation for the failed `pay`?

We were currently planning to expire these on invoice expiry, but on second thought it would be useful to be able to withdraw a preapproval for a payment we give up on.  I suppose it should only succeed in removal if there are no pending HTLCs involving that payment hash.

> 
> 
> On a more design-related thing: I'm assuming that `keysend` is the reason why we expose the pre-approval on the JSON-RPC. It being a plugin, it can't ask the signer to pre-approve, is that correct?

I'll defer to @ksedgwic on this question.

**Comment by ksedgwic** [comment]
> On a more design-related thing: I'm assuming that keysend is the reason why we expose the pre-approval on the JSON-RPC. It being a plugin, it can't ask the signer to pre-approve, is that correct?

Yes, that's right

**Comment by ksedgwic** [comment]
> Would pre-approval result in any state changes on the signer? I.e., does the signer remember it approved a send, and would consider that for other decisions as well? Is this persisted?

Yes, the approval is remembered and persisted.  Subsequent preapproval requests for an approved invoice or keysend are also immediately accepted.

File: lightningd/invoice.c
Comment (cdecker) [inline lightningd/invoice.c:1830]: TIL; I did not know we accept these prefixes.

File: lightningd/invoice.c
Comment (cdecker) [inline lightningd/invoice.c:1834]: Doing synchronous comms here may not be what we want. It ends up blocking the entire `lightningd` until the signer has responded. This means that if we have a long RTT from `lightningd` to `hsmd` we are unresponsive for that roundtrip. Worse, if the signer does not respond we are blocked forever and can't react to on-chain events and peer messsages.

We should always prefer async comms in `lightningd`.

File: lightningd/invoice.c
Comment (cdecker) [inline lightningd/invoice.c:1878]: Same thing about sync comms above.

**Review by cdecker — CHANGES_REQUESTED** [review]
Just a minor thing about the dangers of sync comms when the outcome itself is not critical itself. It's a relatively new consideration, and therefore not consistently applied to the existing code. I'd be more at ease with async comms :-)

**Comment by cdecker** [comment]
Hi @ksedgwic, sorry for suggesting this without first checking if we do any async comms with the signer. It turns out we don't, and introducing it in this PR would massively expand the scope of this PR. I therefore retract my previous comments, and think the changes should go in as they are.

ACK https://github.com/ElementsProject/lightning/pull/5821/commits/31da26cf2ed053e627982fc21858067117b333eb

**Comment by cdecker** [comment]
@endothermicdev would you like me to rebase this to get rid of the `kick ci` commit at the end? It's not semantically meaningful and was there solely to convince CI to give it another shot.

@ksedgwic if CI gets stuck again I can kick it on the console :-)

**Comment by endothermicdev** [comment]
@cdecker  yes, I would appreciate that.  Thanks for the updating your review as well.

**Comment by cdecker** [comment]
@ksedgwic it seems like the PR disallows maintainers to push to your branch, can you rebase and drop the last commit? Then we'll fold the PR into the release 🎉

**Comment by devrandom** [comment]
I can't change that flag from here, but I was able to rebase and drop the extra commit.

**Comment by devrandom** [comment]
CI fail seems unrelated...

**Comment by endothermicdev** [comment]
Yes, unrelated but I haven't identified the fix to that one yet.  Thanks!

ACK 0fa933e

**Pull request merged**
